third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerAref.cpp

```diff
@@ -331,11 +337,18 @@ LogicalResult rewriteGetEnterOp(ArefCreateOp arefOp, ArefGetEnterOp op,
 }
 
 LogicalResult insertArriveBarrier(Location loc, ArrayAttr asyncOps,
-                                  PatternRewriter &rewriter, Value mbar) {
-
+                                  PatternRewriter &rewriter, Value mbar,
+                                  int count) {
   for (auto asyncOp : asyncOps) {
     auto asyncOpEnum = cast<AsyncOpAttr>(asyncOp).getValue();
-    rewriter.create<nvws::AsyncCompleteOp>(loc, mbar, asyncOpEnum);
+    switch (asyncOpEnum) {
+    case AsyncOp::NONE:
+      rewriter.create<nvidia_gpu::ArriveBarrierOp>(loc, mbar, count);
```

Comment:
To check my understanding, we need this WA because we do depend on a bar.sync and a single-thread arrive to avoid the issue with `OptimizePartitionWarps`?

And resolving this issue should improve perf?

Reply:
> To check my understanding, we need this WA because we do depend on a bar.sync and a single-thread arrive to avoid the issue with OptimizePartitionWarps?

correct

> And resolving this issue should improve perf?
data seems to indicate this, but I'm rerunning with locked clock  and larger warmp/rep count.